features/include/pcl/features/impl/good.hpp

```diff
@@ -0,0 +1,666 @@
+  * 
+  * \note This is an implementation of the GOOD descriptor which has been presented in the following papers:
+  * 
+  *	[1] Kasaei, S. Hamidreza,  Ana Maria Tomé, Luís Seabra Lopes, Miguel Oliveira 
+  *	"GOOD: A global orthographic object descriptor for 3D object recognition and manipulation." 
+  *	Pattern Recognition Letters 83 (2016): 312-320.http://dx.doi.org/10.1016/j.patrec.2016.07.006
+  *
+  *	[2] Kasaei, S. Hamidreza, Luís Seabra Lopes, Ana Maria Tomé, Miguel Oliveira 
+  * 	"An orthographic descriptor for 3D object learning and recognition." 
+  *	2016 IEEE/RSJ International Conference on Intelligent Robots and Systems (IROS), Daejeon, 2016, 
+  *	pp. 4158-4163. doi: 10.1109/IROS.2016.7759612
+  * 
+  * Please adequately refer to this work any time this code is being used by citing above papers.
+  * If you do publish a paper where GOOD descriptor helped your research, we encourage you to cite the above papers in your publications.
+  * 
+  * \author Hamidreza Kasaei (Seyed.Hamidreza[at]ua[dot]pt  Kasaei.Hamidreza[at]gmail[dot]com )
+  */
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void 
+pcl::GOODEstimation<PointInT, NumberOfBins>::computeBoundingBoxDimensions (PointCloudIn pc, pcl::PointXYZ &dimensions)
+{
+  PointInT minimum_pt;
+  PointInT maximum_pt;
+  
+  pcl::getMinMax3D (*pc, minimum_pt, maximum_pt); // min max for bounding box
+  dimensions.x = (maximum_pt.x - minimum_pt.x); 
+  dimensions.y = (maximum_pt.y - minimum_pt.y); 
+  dimensions.z = (maximum_pt.z - minimum_pt.z); 	
+  //NOTE: I changed the above lines as follow, but I faced several error.
+  //dimensions = maximum_pt.getVector3fMap () - minimum_pt.getVector3fMap ();
+
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void
+pcl::GOODEstimation<PointInT, NumberOfBins>::projectPointCloudToPlane (PointCloudIn pc_in, boost::shared_ptr<pcl::ModelCoefficients> coefficients, PointCloudIn pc_out)
+{
+  //Create the projection object
+  pcl::ProjectInliers<PointInT> projection;
+  projection.setModelType (pcl::SACMODEL_NORMAL_PLANE); //set model type
+  projection.setInputCloud (pc_in);
+  projection.setModelCoefficients (coefficients);
+  projection.filter (*pc_out);
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void 
+pcl::GOODEstimation<PointInT, NumberOfBins>::convert2DHistogramTo1DHistogram (std::vector<std::vector<unsigned int> >  histogram_2D, std::vector<unsigned int>  &histogram)
+{
+  for (size_t i = 0 ; i < histogram_2D.size(); i++)
+  {   
+    for (size_t j = 0; j < histogram_2D.at(i).size(); j++)
+    {
+      histogram.push_back (histogram_2D.at(i).at(j));
+    }
+  }
+}
```

Comment:
`std::array` is C++11, we can not use it just now. What we can use is `Eigen::Array` with fixed size, either 1 or 2 dimensional.

@SergioRAgostinho don't get your idea, please elaborate.